File: python_prototype/units.py
# units.py
#
# This file defines:
#   - The Unit class (one object per counter)
#   - A factory function that creates every counter in the game
#
# Each Unit stores all the data printed on a physical game counter.


class Unit:
    """
    Represents one physical game counter.

    Attributes
    ----------
    name        : str   — The counter's label (e.g. "Rusich")
    side        : str   — "Wagner" or "Russia"
    base_sp     : int   — Strength Points printed on the counter
    base_mp     : int   — Movement Points printed (0 = Wagner, uses MCT instead)
    reduced     : bool  — Does this counter have a reduced (flip) side?
    is_reduced  : bool  — Is it currently flipped to the reduced side?
    switchable  : bool  — (Z) Can this unit switch sides? (future feature)
    police      : bool  — (P) Cannot initiate combat (future feature)
    in_cup      : bool  — (C) Deploys via mobilization cup (future feature)
    location    : str   — Current map location, or None if off-map
    mp_spent    : int   — MP used so far this turn (reset each turn)
    """

    def __init__(
        self,
        name,
        side,
        base_sp,
        base_mp=0,
        reduced=False,
        switchable=False,
        police=False,
        in_cup=False,
        location=None,
    ):
        self.name       = name
        self.side       = side
        self.base_sp    = base_sp
        self.base_mp    = base_mp
        self.reduced    = reduced       # HAS a reduced side
        self.is_reduced = False         # Currently on reduced side?
        self.switchable = switchable
        self.police     = police
        self.in_cup     = in_cup
        self.location   = location
        self.mp_spent   = 0            # Tracks movement this turn

    @property
    def current_sp(self):
        """
        Current Strength Points.
        If the unit is flipped to its reduced side, SP drops by 1 (minimum 1).
        """
        if self.is_reduced and self.reduced:
            return max(1, self.base_sp - 1)
        return self.base_sp

    def is_wagner(self):
        """Convenience check — Wagner units get MP from MCT, not base_mp."""
        return self.side == "Wagner"

    def is_helicopter(self):
        """Helicopters ignore river crossing penalties."""
        return self.name == "Helicopters"

    def reset_mp(self):
        """Call at the start of each turn to reset movement spending."""
        self.mp_spent = 0

    def __repr__(self):
        """Human-readable one-line summary of this unit."""
        loc_str     = self.location if self.location else "Off Map"
        reduced_str = " [REDUCED]" if self.is_reduced else ""
        return (
            f"[{self.side}] {self.name}"
            f" | SP:{self.current_sp} MP:{self.base_mp}"
            f" | {loc_str}{reduced_str}"
        )


# ---------------------------------------------------------------------------
# Factory — builds every counter described in the PMJ Counter Manifest
# ---------------------------------------------------------------------------

def create_all_units():
    """
    Create and return a list of every Unit in the game.

    Starting locations:
      - Wagner units begin at Rostov-On-Don (the march's starting point).
      - Russian units begin off-map (location=None) until placed.
    """
    units = []

    # ── Wagner ──────────────────────────────────────────────────────────────
    # Wagner units have base_mp=0 because their MP is set by the MCT.
    units.append(Unit(
        name="Rusich", side="Wagner", base_sp=1,
        location="Rostov-On-Don"
    ))
    units.append(Unit(
        name="Utkin", side="Wagner", base_sp=2,
        location="Rostov-On-Don"
    ))
    units.append(Unit(
        name="Serb", side="Wagner", base_sp=1,
        location="Rostov-On-Don"
    ))

    # ── Russia ───────────────────────────────────────────────────────────────
    units.append(Unit(
        name="Mechanized Regiment", side="Russia",
        base_sp=3, base_mp=4, switchable=True, in_cup=True,
    ))
    units.append(Unit(
        name="Motorized Infantry", side="Russia",
        base_sp=1, base_mp=3, switchable=True, in_cup=True
    ))
    units.append(Unit(
        name="Armored Regiment", side="Russia",
        base_sp=3, base_mp=3, switchable=True, in_cup=True
    ))
    units.append(Unit(
        name="OMON", side="Russia",
        base_sp=2, base_mp=2, reduced=True, police=True,
        location="Moscow"
    ))
    units.append(Unit(
        name="Akhmat", side="Russia",
        base_sp=2, base_mp=2, reduced=True, in_cup=True
    ))
    units.append(Unit(
        name="FSB", side="Russia",
        base_sp=2, base_mp=2, reduced=True,
        location="Moscow"
    ))
    units.append(Unit(
        name="Helicopters", side="Russia",
        base_sp=3, base_mp=4, reduced=True, in_cup=True
    ))
    units.append(Unit(
        name="SOBR", side="Russia",
        base_sp=2, base_mp=2, reduced=True,
        location="Moscow"
    ))
    units.append(Unit(
        name="MOSpol", side="Russia",
        base_sp=1, base_mp=2, police=True, in_cup=True
    ))

    return units

File: pmj/src/units.rs
/// Units module — all game counters from the PMJ Counter Manifest.

use crate::map::Location;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Side {
    Wagner,
    Russia,
}

/// Unique identifier for each unit type in the game.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum UnitId {
    // Wagner
    Rusich,
    Utkin,
    Serb,
    // Russia
    MechanizedRegiment,
    MotorizedInfantry,
    ArmoredRegiment,
    Omon,
    Akhmat,
    Fsb,
    Helicopters,
    Sobr,
    Mospol,
}

impl UnitId {
    pub fn name(&self) -> &'static str {
        match self {
            UnitId::Rusich => "Rusich",
            UnitId::Utkin => "Utkin",
            UnitId::Serb => "Serb",
            UnitId::MechanizedRegiment => "Mech Regiment",
            UnitId::MotorizedInfantry => "Motor Infantry",
            UnitId::ArmoredRegiment => "Armor Regiment",
            UnitId::Omon => "OMON",
            UnitId::Akhmat => "Akhmat",
            UnitId::Fsb => "FSB",
            UnitId::Helicopters => "Helicopters",
            UnitId::Sobr => "SOBR",
            UnitId::Mospol => "MOSpol",
        }
    }

    /// Short label for compact display (max 4 chars).
    pub fn short(&self) -> &'static str {
        match self {
            UnitId::Rusich => "RSCH",
            UnitId::Utkin => "UTKN",
            UnitId::Serb => "SERB",
            UnitId::MechanizedRegiment => "MECH",
            UnitId::MotorizedInfantry => "MOTR",
            UnitId::ArmoredRegiment => "ARMR",
            UnitId::Omon => "OMON",
            UnitId::Akhmat => "AKHM",
            UnitId::Fsb => "FSB ",
            UnitId::Helicopters => "HELI",
            UnitId::Sobr => "SOBR",
            UnitId::Mospol => "MPOL",
        }
    }

    /// NATO-style center symbol for halfblock counter display.
    pub fn nato_symbol(&self) -> char {
        match self {
            // Wagner: use starting letter for visual distinction
            UnitId::Rusich => 'R',
            UnitId::Utkin => 'U',
            UnitId::Serb => 'S',
            // Russian military: NATO infantry symbol
            UnitId::MechanizedRegiment => '╳',
            UnitId::MotorizedInfantry => '╳',
            UnitId::ArmoredRegiment => '◇',
            UnitId::Omon => '╳',
            UnitId::Akhmat => '╳',
            UnitId::Fsb => '█',
            UnitId::Helicopters => 'H',
            UnitId::Sobr => '█',
            UnitId::Mospol => '╳',
        }
    }

    pub fn is_wagner(&self) -> bool {
        matches!(self, UnitId::Rusich | UnitId::Utkin | UnitId::Serb)
    }

    /// All Wagner unit IDs.
    pub fn wagner_units() -> &'static [UnitId] {
        &[UnitId::Rusich, UnitId::Utkin, UnitId::Serb]
    }
}

/// A single game counter.
#[derive(Debug, Clone)]
pub struct Unit {
    pub id: UnitId,
    pub side: Side,
    pub base_sp: i32,
    pub base_mp: i32,
    pub has_reduced_side: bool,
    pub is_reduced: bool,
    pub switchable: bool,  // (Z) can switch sides Russia -> Wagner
    pub police: bool,      // (P) no offensive capability
    pub in_cup: bool,      // (C) starts in Moscow Mobilization Cup
    pub location: Option<Location>,
    pub mp_spent: i32,
    pub dispersed: bool,
}

impl Unit {
    /// Current SP accounting for reduction (reduced = base_sp - 1, min 1).
    pub fn current_sp(&self) -> i32 {
        if self.is_reduced && self.has_reduced_side {
            (self.base_sp - 1).max(1)
        } else {
            self.base_sp
        }
    }

    pub fn is_wagner(&self) -> bool {
        self.side == Side::Wagner
    }

    pub fn is_helicopter(&self) -> bool {
        self.id == UnitId::Helicopters
    }

    pub fn is_on_map(&self) -> bool {
        self.location.is_some()
    }

    pub fn reset_mp(&mut self) {
        self.mp_spent = 0;
    }

    /// Step-reduce this unit. Returns true if reduced, false if eliminated.
    pub fn step_reduce(&mut self) -> bool {
        if self.has_reduced_side && !self.is_reduced {
            self.is_reduced = true;
            true
        } else {
            self.eliminate();
            false
        }
    }

    /// Remove unit from the map.
    pub fn eliminate(&mut self) {
        self.location = None;
        self.is_reduced = false;
        self.dispersed = false;
    }
}

/// Create all units per the PMJ Counter Manifest.
pub fn create_all_units() -> Vec<Unit> {
    vec![
        // === Wagner ===
        // Wagner units have base_mp=0 because MP comes from MCT
        Unit {
            id: UnitId::Rusich,
            side: Side::Wagner,
            base_sp: 1,
            base_mp: 0,
            has_reduced_side: false,
            is_reduced: false,
            switchable: false,
            police: false,
            in_cup: false,
            location: Some(Location::RostovOnDon),
            mp_spent: 0,
            dispersed: false,
        },
        Unit {
            id: UnitId::Utkin,
            side: Side::Wagner,
            base_sp: 2,
            base_mp: 0,
            has_reduced_side: false,
            is_reduced: false,
            switchable: false,
            police: false,
            in_cup: false,
            location: Some(Location::RostovOnDon),
            mp_spent: 0,
            dispersed: false,
        },
        Unit {
            id: UnitId::Serb,
            side: Side::Wagner,
            base_sp: 1,
            base_mp: 0,
            has_reduced_side: false,
            is_reduced: false,
            switchable: false,
            police: false,
            in_cup: false,
            location: Some(Location::RostovOnDon),
            mp_spent: 0,
            dispersed: false,
        },
        // === Russia ===
        Unit {
            id: UnitId::MechanizedRegiment,
            side: Side::Russia,
            base_sp: 3,
            base_mp: 4,
            has_reduced_side: false,
            is_reduced: false,
            switchable: true,
            police: false,
            in_cup: true,
            location: None,
            mp_spent: 0,
            dispersed: false,
        },
        Unit {
            id: UnitId::MotorizedInfantry,
            side: Side::Russia,
            base_sp: 1,
            base_mp: 3,
            has_reduced_side: false,
            is_reduced: false,
            switchable: true,
            police: false,
            in_cup: true,
            location: None,
            mp_spent: 0,
            dispersed: false,
        },
        Unit {
            id: UnitId::ArmoredRegiment,
            side: Side::Russia,
            base_sp: 3,
            base_mp: 3,
            has_reduced_side: false,
            is_reduced: false,
            switchable: true,
            police: false,
            in_cup: true,
            location: None,
            mp_spent: 0,
            dispersed: false,
        },
        Unit {
            id: UnitId::Omon,
            side: Side::Russia,
            base_sp: 2,
            base_mp: 2,
            has_reduced_side: true,
            is_reduced: false,
            switchable: false,
            police: true,
            in_cup: false,
            location: Some(Location::Moscow),
            mp_spent: 0,
            dispersed: false,
        },
        Unit {
            id: UnitId::Akhmat,
            side: Side::Russia,
            base_sp: 2,
            base_mp: 2,
            has_reduced_side: true,
            is_reduced: false,
            switchable: false,
            police: false,
            in_cup: true,
            location: None,
            mp_spent: 0,
            dispersed: false,
        },
        Unit {
            id: UnitId::Fsb,
            side: Side::Russia,
            base_sp: 2,
            base_mp: 2,
            has_reduced_side: true,
            is_reduced: false,
            switchable: false,
            police: false,
            in_cup: false,
            location: Some(Location::Moscow),
            mp_spent: 0,
            dispersed: false,
        },
        Unit {
            id: UnitId::Helicopters,
            side: Side::Russia,
            base_sp: 3,
            base_mp: 4,
            has_reduced_side: true,
            is_reduced: false,
            switchable: false,
            police: false,
            in_cup: true,
            location: None,
            mp_spent: 0,
            dispersed: false,
        },
        Unit {
            id: UnitId::Sobr,
            side: Side::Russia,
            base_sp: 2,
            base_mp: 2,
            has_reduced_side: true,
            is_reduced: false,
            switchable: false,
            police: false,
            in_cup: false,
            location: Some(Location::Moscow),
            mp_spent: 0,
            dispersed: false,
        },
        Unit {
            id: UnitId::Mospol,
            side: Side::Russia,
            base_sp: 1,
            base_mp: 2,
            has_reduced_side: false,
            is_reduced: false,
            switchable: false,
            police: true,
            in_cup: true,
            location: None,
            mp_spent: 0,
            dispersed: false,
        },
    ]
}
